[Merged by Bors] - chore(algebra/group/type_tags): Use ≃ instead of →

eric-wieser opened the pull request:
These functions are all equivalences, so we may as well expose that in their type

This also fills in some conversions that were missing.

Unfortunately this requires some type ascriptions in a handful of places.
It might be possible to remove these somehow.
This zulip thread contains a mwe: https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Type.20inference.20on.20.60.E2.86.92.60.20vs.20.60.E2.89.83.60/near/211922501.

---
<!-- put comments you want to keep out of the PR commit here -->

eric-wieser committed:
chore(algebra/group/type_tags): Use ≃ instead of →

These functions are all equivalences, so we may as well expose that in their type.
This effectively renames `of_add_injective` to `multiplicative.of_add.injective`, and the same for the other three lemmas.

Vierkantor reviewed (approved):
I'm curious whether any expressions of the form `(_ : additive _).to_mul` will break. It seems `coe_fn` has become a lot more robust recently (making #4345 a lot more possible).

eric-wieser committed:
chore(dynamics/circle/rotation_number/translation_number): Fix type resolution issues

eric-wieser committed:
chore(data/equiv/mul_add): Use ≃ instead of →

These functions are all equivalences, so we may as well expose that in their type.

This also fills in some equivalences that were missing.

eric-wieser commented on src/algebra/group/type_tags.lean:
```diff
@@ -166,21 +161,32 @@ instance [comm_group α] : add_comm_group (additive α) :=
-def monoid_hom.to_additive' [monoid α] [add_monoid β] (f : α →* multiplicative β) :
-  additive α →+ β :=
-⟨f.1, f.2, f.3⟩
+/-- Reinterpret `α →+ β` as `multiplicative α →* multiplicative β`. -/
+def add_monoid_hom.to_multiplicative [add_monoid α] [add_monoid β] :
+  (α →+ β) ≃ (multiplicative α →* multiplicative β) :=
+⟨λ f, ⟨f.1, f.2, f.3⟩, λ f, ⟨f.1, f.2, f.3⟩, λ x, by { ext, refl, }, λ x, by { ext, refl, }⟩
```
These docstrings seem stupid, because they say nothing other than what the type already says. Should I remove them and use `nolint` instead?

Vierkantor replied:
I'm weakly in favour of keeping it as is, at least it indicates the map does the right thing (instead of say, sending everything to `1`).

eric-wieser replied:
Well, a map that sent everything to `1` wouldn't be a `≃` ;)

Vierkantor replied:
Fair enough :P

eric-wieser replied:
I'm happy to leave this as is though, rather that waiting for another full CI run.

eric-wieser commented on src/algebra/group/type_tags.lean:
```diff
@@ -166,21 +161,32 @@ instance [comm_group α] : add_comm_group (additive α) :=
+/-- Reinterpret `α →+ additive β` as `multiplicative α →* β`. -/
+def add_monoid_hom.to_multiplicative'' [add_monoid α] [monoid β] :
+  (α →+ additive β) ≃ (multiplicative α →* β) :=
+⟨λ f, ⟨f.1, f.2, f.3⟩, λ f, ⟨f.1, f.2, f.3⟩, λ x, by { ext, refl, }, λ x, by { ext, refl, }⟩
```
A lot of the `by { ext, refl, }`s could be replaced with `add_monoid_hom.ext (λ _, rfl)` but I don't know if it is worth it.

Vierkantor replied:
I'm happy with it either way. The tactic proof omits a little bit of noise, since it doesn't need the `add_monoid_hom` qualification. Since it's such a short proof, that's not really important though.

eric-wieser replied:
I'll leave it as is then.

jcommelin reviewed (approved):
Thanks :tada:

bors merge

eric-wieser committed:
chore(ring_theory/roots_of_unity): Add explicit casts
